package/base-files/files/lib/functions/system.sh

```diff
@@ -1,31 +1,36 @@
 # Copyright (C) 2006-2013 OpenWrt.org
 
 . /lib/functions.sh
-. /usr/share/libubox/jshn.sh
 
 get_mac_binary() {
 	local path="$1"
-	local offset="$2"
+	local offset="${2:-0}"
+	local length="${3:-6}"
+	local mac
 
 	if ! [ -e "$path" ]; then
 		echo "get_mac_binary: file $path not found!" >&2
 		return
 	fi
 
-	hexdump -v -n 6 -s $offset -e '5/1 "%02x:" 1/1 "%02x"' $path 2>/dev/null
+	mac=$(head -c $((offset + length)) "$path" | tail -c "$length" |
+		hexdump -e '6/1 "%02x"' 2>/dev/null)
+
+	macaddr_canonicalize "$mac"
 }
 
 get_mac_label_dt() {
 	local basepath="/proc/device-tree"
-	local macdevice="$(cat "$basepath/aliases/label-mac-device" 2>/dev/null)"
-	local macaddr
+	local macdevice macaddr
 
+	macdevice=$(cat "$basepath/aliases/label-mac-device" 2>/dev/null)
 	[ -n "$macdevice" ] || return
 
-	macaddr=$(get_mac_binary "$basepath/$macdevice/mac-address" 0 2>/dev/null)
-	[ -n "$macaddr" ] || macaddr=$(get_mac_binary "$basepath/$macdevice/local-mac-address" 0 2>/dev/null)
+	[ -e "$basepath/$macdevice/mac-address" ] &&
+		macaddr=$(get_mac_binary "$basepath/$macdevice/mac-address" 2>/dev/null) ||
+		macaddr=$(get_mac_binary "$basepath/$macdevice/local-mac-address" 2>/dev/null)
 
```

Comment:
Strictly speaking this changes logic. If `get_mac_binary()` would result an empty result in the old code (e.g. while reading a zero-length file), the fallback logic would be used. Now it decides solely on the presence of the file (and introduces a redundant stat syscall instead of a simple string length check)